src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonMetadataServicesConverter.cs

```diff
@@ -15,44 +15,20 @@ namespace System.Text.Json.Serialization.Converters
     /// <typeparam name="T">The type to converter</typeparam>
     internal sealed class JsonMetadataServicesConverter<T> : JsonResumableConverter<T>
     {
-        private readonly Func<JsonConverter<T>>? _converterCreator;
-        private JsonConverter<T>? _converter;
-
         // A backing converter for when fast-path logic cannot be used.
-        internal JsonConverter<T> Converter
-        {
-            get
-            {
-                _converter ??= _converterCreator!();
-                Debug.Assert(_converter != null);
-                Debug.Assert(_converter.ConverterStrategy == ConverterStrategy);
-                return _converter;
-            }
-        }
+        internal JsonConverter<T> Converter { get; }
 
         internal override Type? KeyType => Converter.KeyType;
-
         internal override Type? ElementType => Converter.ElementType;
 
         internal override bool ConstructorIsParameterized => Converter.ConstructorIsParameterized;
         internal override bool SupportsCreateObjectDelegate => Converter.SupportsCreateObjectDelegate;
         internal override bool CanHaveMetadata => Converter.CanHaveMetadata;
 
-        public JsonMetadataServicesConverter(Func<JsonConverter<T>> converterCreator, ConverterStrategy converterStrategy)
```

Comment:
It was written this way so that we aren't initializing converters that won't be used at runtime (due to fast-path). Start-up perf might be affected.

Reply:
I figured this might have been the initial motivation, but ultimately this ends up allocating an intermediate `Func<T>` which more or less costs the same as allocating your typical converter. There's also the added cost of lazy initialization, dynamic dispatch when invoking the delegate and the fact that the delegate is being stored even after initialization has been completed.

Reply:
> which more or less costs the same as allocating your typical converter

For my education how do you determine this?

Reply:
Unless you have a particularly large object or array, the overheads of allocating, storing and collecting an object in the managed heap are effectively the same.

There are some nice details in this series: https://devblogs.microsoft.com/premier-developer/managed-object-internals-part-4-fields-layout/